File: iOS7bookExamples/bk2ch06p347SimplestRestorationExample2/ch19p652SimplestRestorationExample2/SecondViewController.m


#import "SecondViewController.h"
#import "PresentedViewController.h"

@interface SecondViewController () <UIViewControllerRestoration>

@end

@implementation SecondViewController

-(void)encodeRestorableStateWithCoder:(NSCoder *)coder {
    [super encodeRestorableStateWithCoder:coder];
    NSLog(@"%@ encode %@", self, coder);
}

-(void)decodeRestorableStateWithCoder:(NSCoder *)coder {
    [super decodeRestorableStateWithCoder:coder];
    NSLog(@"%@ decode %@", self, coder);
}

-(void)applicationFinishedRestoringState {
    NSLog(@"finished %@", self);
}


-(void)viewDidLoad {
    [super viewDidLoad];
    NSLog(@"view did load %@", self);
    self.view.backgroundColor = [UIColor yellowColor];
    UIButton* button = [UIButton buttonWithType:UIButtonTypeRoundedRect];
    [button setTitle:@"Present" forState:UIControlStateNormal];
    [button addTarget:self
               action:@selector(doPresent:)
     forControlEvents:UIControlEventTouchUpInside];
    [button sizeToFit];
    button.center = self.view.center;
    [self.view addSubview:button];
}

+ (UIViewController*) makePresentedViewController {
    PresentedViewController* pvc = [PresentedViewController new];
    pvc.restorationIdentifier = @"presented";
    pvc.restorationClass = [self class];
    return pvc;
}

+ (UIViewController*) viewControllerWithRestorationIdentifierPath:(NSArray*)ip
                                                            coder:(NSCoder*)coder {
    NSLog(@"vcwithrip %@ %@ %@", self, ip, coder);
    UIViewController* vc = nil;
    if ([[ip lastObject] isEqualToString:@"presented"]) {
        vc = [self makePresentedViewController];
    }
    return vc;
}

-(void)doPresent:(id)sender {
    UIViewController* pvc = [[self class] makePresentedViewController];
    [self presentViewController:pvc animated:YES completion:nil];
}



@end

File: iOS8bookExamples/bk2ch06p347SimplestRestorationExample2/ch19p652SimplestRestorationExample2/SecondViewController.swift


import UIKit

class SecondViewController : UIViewController {
    
    
    override func encodeRestorableStateWithCoder(coder: NSCoder) {
        println("\(self) encode \(coder)")
    }
    
    override func decodeRestorableStateWithCoder(coder: NSCoder) {
        println("\(self) decode \(coder)")
    }
    
    override func applicationFinishedRestoringState() {
        println("finished \(self)")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        println("view did load \(self)")
        self.view.backgroundColor = UIColor.yellowColor()
        let button = UIButton.buttonWithType(.System) as! UIButton
        button.setTitle("Present", forState:.Normal)
        button.addTarget(self,
            action:"doPresent:",
            forControlEvents:.TouchUpInside)
        button.sizeToFit()
        button.center = self.view.center
        self.view.addSubview(button)
    }
    
    class func makePresentedViewController () -> UIViewController {
        let pvc = PresentedViewController()
        pvc.restorationIdentifier = "presented"
        pvc.restorationClass = self
        return pvc
    }
    
    func doPresent(sender:AnyObject?) {
        let pvc = self.dynamicType.makePresentedViewController()
        self.presentViewController(pvc, animated:true, completion:nil)
    }
}

// a classic mistake is to implement viewControllerWithRestorationIdentifierPath...
// but forget to declare explicit conformance to UIViewControllerRestoration
// in that case, restoration will fail, without a crash,
// but now (new in iOS 8?) with a delightfully helpful message:
// "Warning: restoration class for view controller does not conform to UIViewControllerRestoration protocol: Class is ..."

extension SecondViewController : UIViewControllerRestoration {
    class func viewControllerWithRestorationIdentifierPath(ip: [AnyObject],
        coder: NSCoder) -> UIViewController? {
            println("vcwithrip \(NSStringFromClass(self)) \(ip) \(coder)")
            var vc : UIViewController? = nil
            let last = ip.last as! String
            switch last {
            case "presented":
                vc = self.makePresentedViewController()
            default: break
            }
            return vc
    }
}

